### PR TITLE
zoneconcierge: permissioned integration

### PR DESCRIPTION
**Pull request by SebastianElvis**
Resolves https://github.com/babylonlabs-io/pm/issues/44

This PR implements the permissioned integration and generic consumer registry, including

- [x] generic consumer registry compatible with Cosmos/ETH L2 consumers
- [x] parameter for governance gating consumer addition
- [x] ensuring IBC channel creation only when the IBC light client is in consumer registry
- [x] removing the Cosmos consumer registration upon IBC handshake in Babylon
- [x] removing the Cosmos consumer registration upon IBC handshake in Babylon contract
- [x] fuzz tests for Cosmos integration
- [x] e2e tests for Cosmos integration

ADR: https://github.com/babylonlabs-io/pm/pull/121
Accompanying PR in the contract side: https://github.com/babylonlabs-io/babylon-contract/pull/87

There are two TODOs that will be addressed in subsequent PRs:

- msg for removing registered consumer
- fuzz/e2e test for ETH L2 integrations

### CODE REVIEW

File: proto/babylon/btcstkconsumer/v1/btcstkconsumer.proto
Comment (maurolacy) [inline proto/babylon/btcstkconsumer/v1/btcstkconsumer.proto:23]: Nit: What about adding this field to both metadatas below, with a proper description for each?

File: proto/babylon/btcstkconsumer/v1/btcstkconsumer.proto
Comment (maurolacy) [inline proto/babylon/btcstkconsumer/v1/btcstkconsumer.proto:39]: This is redundant with the main struct. Better to move `consumer_id` here, and replicate it in the Ethereum struct, with a proper description.

Moreover: Perhaps it's simpler / clearer to just make `ConsumerRegister` a `oneof`. At the expense of some redundancy in the consumer name and description fields.

File: proto/babylon/btcstkconsumer/v1/params.proto
Comment (maurolacy) [inline proto/babylon/btcstkconsumer/v1/params.proto:16]: Can we set a default here? Just curious.

File: proto/babylon/btcstkconsumer/v1/tx.proto
Comment (maurolacy) [inline proto/babylon/btcstkconsumer/v1/tx.proto:57]: This is a little obscure. This is the Cosmos address of the "assistant" finality contract in the Babylon chain for that Eth-L2 consumer, right? Better to clarify all that in the comments.

Again, what about making `MsgRegisterConsumer` a `oneof` here? Not sure this is a good idea, or even possible, just saying.

File: proto/babylon/zoneconcierge/v1/packet.proto
Comment (maurolacy) [inline proto/babylon/zoneconcierge/v1/packet.proto:21]: Bye nice automatic consumer registration.🙂

File: proto/babylon/zoneconcierge/v1/packet.proto
Comment (maurolacy) [inline proto/babylon/zoneconcierge/v1/packet.proto:21]: Nit: Again, changing the tags is not strictly needed.

**Review by maurolacy — COMMENTED** [review]
Partial review of the build targets / scripts and proto files.

Will review the code next.

File: Makefile
Comment (SebastianElvis) [reply Makefile:280]: I was basically following other e2e entries, but yes you are right that the target should not have dependencies. Added it back and renamed this entry to `test-e2e-bcd-consumer-integration`

File: proto/babylon/btcstaking/v1/packet.proto
Comment (SebastianElvis) [reply proto/babylon/btcstaking/v1/packet.proto:61]: Yep let's have clean proto files at least before the phase-3 testnet/mainnet launch

File: proto/babylon/btcstkconsumer/v1/btcstkconsumer.proto
Comment (SebastianElvis) [reply proto/babylon/btcstkconsumer/v1/btcstkconsumer.proto:23]: I would prefer having it here because every consumer, regardless of its type, will have a consumer ID. Wdyt?

File: proto/babylon/btcstkconsumer/v1/btcstkconsumer.proto
Comment (SebastianElvis) [reply proto/babylon/btcstkconsumer/v1/btcstkconsumer.proto:39]: Yep this is redundant, let's remove this field. 

Personally I still prerfer to keep the current structure, as there might exist some applications that only care about id/name/description. We can follow up on this after backend/frontend comes up with concrete requirements.

File: proto/babylon/btcstkconsumer/v1/params.proto
Comment (SebastianElvis) [reply proto/babylon/btcstkconsumer/v1/params.proto:16]: It's set [here](https://github.com/babylonlabs-io/babylon/blob/de84d8817da424189fa5110ac00c2e3dc1333949/x/btcstkconsumer/types/params.go#L17-L18)

File: proto/babylon/zoneconcierge/v1/packet.proto
Comment (SebastianElvis) [reply proto/babylon/zoneconcierge/v1/packet.proto:21]: Yeah somehow this turns out to be much more complicated -- both Babylon and consumer have to pass gov prop for integration, and one cannot simply deploy a chain that integrates with Babylon without any approval

File: contrib/images/ibcsim-bcd/Dockerfile
Comment (SebastianElvis) [reply contrib/images/ibcsim-bcd/Dockerfile:58]: Make sense. Let me create a PR for babylon-sdk, and let this Dockerfile use the latest `main` of babylon-sdk. Wdyt?

Comment (SebastianElvis) [reply contrib/images/ibcsim-bcd/Dockerfile:58]: https://github.com/babylonlabs-io/babylon-sdk/pull/65

**Comment by SebastianElvis** [comment]
Thanks @maurolacy for the comments! I have revised all of them and answered some questions. Please feel free to take a look again

TODO before merging

- [ ] merge https://github.com/babylonlabs-io/babylon-sdk/pull/65

File: x/btcstkconsumer/types/expected_keepers.go
Comment (maurolacy) [inline x/btcstkconsumer/types/expected_keepers.go:32]: Why do we need this channel keeper if it provides no interface? Just curious.

File: x/btcstkconsumer/types/params.go
Comment (maurolacy) [inline x/btcstkconsumer/types/params.go:17]: 👍🏼

File: Makefile
Comment (maurolacy) [inline Makefile:280]: So, the issue or trade-off here is that this is not cached anymore?

We can still have a no-deps low level 'cache' target here if we want. The 'no-cache' target then calls the 'cache' option, which is there for convenience / user experience.

File: .github/workflows/ci.yml
Comment (maurolacy) [inline .github/workflows/ci.yml:246]: 🎉

**Review by maurolacy — APPROVED** [review]
Looks good to me. Still have to take a look at tests (~and related PR, next~), but approving to unblock.

File: contrib/images/ibcsim-bcd/Dockerfile
Comment (maurolacy) [inline contrib/images/ibcsim-bcd/Dockerfile:21]: What about releasing `v0.6.0-rc.1`. I know it's not semantically version correct, but better than this, which defeats versioning altogether.

File: x/btcstkconsumer/keeper/msg_server.go
Comment (maurolacy) [inline x/btcstkconsumer/keeper/msg_server.go:64]: Wouldn't this require that the IBC connection is established **before** registering the consumer? What is the right / required sequence for Consumer registration?

File: x/btcstkconsumer/keeper/msg_server_test.go
Comment (maurolacy) [inline x/btcstkconsumer/keeper/msg_server_test.go:98]: At some point it would be good to check that the contract is of a particular type / adheres to some particular API.

File: proto/babylon/zoneconcierge/v1/packet.proto
Comment (maurolacy) [reply proto/babylon/zoneconcierge/v1/packet.proto:21]: On second thought, maybe a bit late for this but, what about falling back to / keeping automatic registration for Cosmos chains, in the case of permissionless integration?

**Comment by maurolacy** [comment]
E2e tests are failing in my local:
```
    --- PASS: TestBCDConsumerIntegrationTestSuite/Test1ChainStartup (2.04s)
    --- PASS: TestBCDConsumerIntegrationTestSuite/Test2RegisterAndIntegrateConsumer (129.87s)
    --- PASS: TestBCDConsumerIntegrationTestSuite/Test3CreateConsumerFinalityProvider (13.45s)
    --- PASS: TestBCDConsumerIntegrationTestSuite/Test4RestakeDelegationToMultipleFPs (123.90s)
    --- FAIL: TestBCDConsumerIntegrationTestSuite/Test5ActivateDelegation (45.31s)
    --- FAIL: TestBCDConsumerIntegrationTestSuite/Test6ConsumerFPRewardsGeneration (60.96s)
    --- PASS: TestBCDConsumerIntegrationTestSuite/Test7BabylonFPCascadedSlashing (21.50s)
    --- FAIL: TestBCDConsumerIntegrationTestSuite/Test8ConsumerFPCascadedSlashing (378.03s)
```

**Comment by maurolacy** [comment]
Retrying the tests. The IBC transfer back timeout should be extended still by five seconds or so.

File: x/btcstaking/keeper/grpc_query.go
Comment (SebastianElvis) [reply x/btcstaking/keeper/grpc_query.go:85]: This was a bug because the query height should be the current block height but not 0

File: test/e2e/bytecode/babylon_contract.wasm
Comment (SebastianElvis) [reply test/e2e/bytecode/babylon_contract.wasm:1]: Probably we will need to prioritise https://github.com/babylonlabs-io/pm/issues/118, i.e., all such e2e tests should be moved to Babylon SDK. Wdyt?

File: x/btcstkconsumer/types/expected_keepers.go
Comment (SebastianElvis) [reply x/btcstkconsumer/types/expected_keepers.go:32]: Good catch, removed channel keeper dependency in btcstkconsumer

File: Makefile
Comment (SebastianElvis) [reply Makefile:280]: Yes the cache was mostly for accelerating GH action, specifically `ci.yml` pipeline.

However, now I find this pipeline conflates a number of things, including unit tests, e2e for BTC staking, e2e for BTC timestamping, e2e for integration, etc.. Probably we will need to split it into multiple ones. After doing that we might not need this cache, because the docker images will only be used for BTC staking integration e2e. Wdyt?

File: contrib/images/ibcsim-bcd/Dockerfile
Comment (SebastianElvis) [reply contrib/images/ibcsim-bcd/Dockerfile:21]: Good idea, let's do this

File: x/btcstkconsumer/keeper/msg_server.go
Comment (SebastianElvis) [reply x/btcstkconsumer/keeper/msg_server.go:64]: Not really, here we are retrieving IBC *light client* but not connection or channel.

So the user story now becomes: creating an IBC lgiht client for consumer -> registering consumer using this client ID -> passing gov prop if permissioned integration is enforced -> creating IBC connection/channel

This is the approach followed by ICS, as discussed in the RFC/ADR

File: x/btcstkconsumer/keeper/msg_server_test.go
Comment (SebastianElvis) [reply x/btcstkconsumer/keeper/msg_server_test.go:98]: Good idea, added

**Pull request merged**
